src/environmentd/src/lib.rs

```diff
@@ -489,19 +489,43 @@ impl Listeners {
         if let Some(catalog_kind) = catalog_kind_impl {
             openable_adapter_storage.set_catalog_kind(catalog_kind);
         }
+
+        let bootstrap_args = BootstrapArgs {
+            default_cluster_replica_size: config.bootstrap_default_cluster_replica_size.clone(),
+            bootstrap_role: config.bootstrap_role,
+        };
+
         let mut adapter_storage = openable_adapter_storage
-            .open(
-                boot_ts,
-                &BootstrapArgs {
-                    default_cluster_replica_size: config
-                        .bootstrap_default_cluster_replica_size
-                        .clone(),
-                    bootstrap_role: config.bootstrap_role,
-                },
-                config.deploy_generation,
-                None,
-            )
+            .open(boot_ts, &bootstrap_args, config.deploy_generation, None)
             .await?;
+
+        // Migrate the storage metadata to the persist-backed catalog. This can
+        // be deleted in the version following this merge.
+        {
+            let stash_factory =
+                mz_stash::StashFactory::from_metrics(Arc::clone(&config.controller.stash_metrics));
+            let tls = mz_tls_util::make_tls(&tokio_postgres::config::Config::from_str(
+                &config.controller.storage_stash_url,
+            )?)?;
+
+            let storage_stash = Box::new(mz_catalog::durable::stash_backed_catalog_state(
+                StashConfig {
+                    stash_factory,
+                    stash_url: config.controller.storage_stash_url.clone(),
+                    schema: None,
+                    tls,
+                },
+            ));
+
+            let mut storage_stash = storage_stash
+                .open(boot_ts, &bootstrap_args, config.deploy_generation, None)
+                .await?;
+
+            adapter_storage
+                .migrate_storage_state_from(&mut storage_stash)
+                .await?;
+        }
```

Comment:
Ok, I think I finally see what's going on. You're using the stash implementation of the `DurableCatalogState` API to open, read, and write the storage controller's `Stash`. I don't think that's a good idea and I'm not totally convinced that it would work. For example `DurableCatalogState` expects certain collections to exist and contain certain keys or it gets very confused.

Instead you should be opening a regular `Stash` (https://github.com/MaterializeInc/materialize/blob/8bc8370afa7950bb320d0568c25e69c90585f520/src/stash/src/postgres.rs#L450-L468) and use that API for reading/writing.

Reply:
At a high level I think this code should look like the following psuedo code (emphasis on psuedo):

```Rust
let stash_factory =
    mz_stash::StashFactory::from_metrics(Arc::clone(&config.controller.stash_metrics));
let tls = mz_tls_util::make_tls(&tokio_postgres::config::Config::from_str(
    &config.controller.storage_stash_url,
)?)?;
let storage_stash: Stash = stash_factory.open(tls);
let storage_stash_txn = storage_stash.transaction();
let adapter_storage_txn = adapter_storage.transaction();
// If there's any storage data in the catalog then the migration must have already happened.
if !adapter_storage_txn.get_storage_data().is_empty() {
    return;
}
adapter_storage_txn.set_storage_data(storage_stash_txn.peek(STORAGE_COLLECTION));
storage_stash_txn.commit();
adapter_storage_txn.commit();
```

Reply:
Yeah--I just needed to use `Stash` directly, rather than trying to turn it into an openable catalog state.